src/backend/distributed/executor/local_executor.c

```diff
@@ -431,6 +407,39 @@ ExtractParametersForLocalExecution(ParamListInfo paramListInfo, Oid **parameterT
 }
 
 
+/*
+ * LocallyExecuteUtilityTask runs a utility command via local execution.
+ */
+static void
+LocallyExecuteUtilityTask(Task *task)
+{
+	/*
+	 * If we roll back to a savepoint, we may no longer be in a query on
+	 * a shard. Reset the value as we go back up the stack.
+	 */
+	uint64 prevInLocalQueryOnShard = InLocalQueryOnShard;
```

Comment:
My understanding was that InLocalQueryOnShard is only set during the command execution. Even if there are nested calls, at the end of the top command finishes, InLocalQueryOnShard is set back to zero.

 I really cannot see why ROLLBACK TO SAVEPOINT should preserve the value. Because, at the start/end of a top-level command, the value should be 0 anyway?

Maybe can you guide me to an example for this?

Reply:
My main goal here was to ensure that if you are in a query on a shard that allows nested execution (e.g. triggers), and then do a nested query on a reference table, we keep the flag as true. However, as soon as we depart the top-level query, we set the flag back to false. This should work correctly even if the function does stuff like throw and catch exceptions (roll back to a savepoint).

Reply:
So, then another equivalent approach could be to reset the flag (to 0) on tx and sub-tx abort callbacks? 

We don't have to do, mostly asking to clarify

Reply:
> So, then another equivalent approach could be to reset the flag (to 0) on tx and sub-tx abort callbacks?

I don't think so. For instance, if a query on a function calls two functions, the first of which has an exception handler that catches an error (e.g. the one we throw here), then should the xact handler reset the flag? Probably not, since we still have the second function call that also needs to be restricted. On the other hand, if we don't reset the flag, then it may be incorrectly set for the next query. The right time to do it is exactly at the end of execution of the query of the shard.

```
-- throws an error in fn1 if it does a query, but fn1 is free to catch that error
SELECT fn1(), fn2() FROM dist WHERE key = 1;
-- allowed
SELECT fn3() FROM ref;
```